### ISSUE TITLE
murder: backend_version() does not recognize 3.1+ which causes downgrade to v6 index

### ISSUE THREAD

**Issue by ensc** [cyrusimap/cyrus-imapd#3123]
`backend_version()` checks for

```c
    if (strstr(be->banner, CYRUS_VERSION)) {
        return MAILBOX_MINOR_VERSION;
    }

    /* master branch? */
    if (strstr(be->banner, "Cyrus IMAP 3.0")) {
        return 13;
    }

    /* version 2.5 is 13 */
   if (strstr(be->banner, "Cyrus IMAP 2.5.")
```

and misses recent 3.1, 3.2 and 3.3 releases.  This causes index downgrade to v6 on XFERs.

Affects all stable versions and legacy 2.4 + 2.5 too.

**Comment by elliefm** [comment]
Good catch.  I'm immediately going to update the comment next to MAILBOX_MINOR_VERSION to note that this must be done when the value changes.  There's already a comment along these lines in `backup_version()` itself, but no-one sees it unless they already know to look there.

**Comment by elliefm** [comment]
While figuring out how to recognise the various 3.x versions, I found what looks like (and I have since confirmed to be) a bug in our exisiting handling of the 2.3.x versions.

We don't have any tests for backend_version(), so I'm down a rabbit hole at the moment making it testable, so that when I do update backend_version() to handle 3.x, I don't make any whoopsies like was made for 2.3.x.

The tests will only be on the master branch, because making it testable changes the libcyrus ABI.  But, once I know the updates to backend_version() work correctly, I'll be able to backport the updates to the older branches even though I won't be able to backport the corresponding tests.

**Comment by elliefm** [comment]
This is now fixed on master, but I still need to backport the relevant bits to the older branches

I think I also want to add another cunit test for nonexistent future versions (3.4 etc) to make sure they're v6, as a prompt for us to remember to also keep the cunit tests up to date...

**Comment by elliefm** [comment]
This is now also fixed on 2.4, 2.5, 3.0, and 3.2 branches.   Need to decide which ones get new releases, and then do the new releases.

**Comment by hagedose** [comment]
I'm trying to understand the implications of this bug. We're still on a standalone 2.4 server, and our current thinking is that the best way forward might be to put a 2.4 frontend in front of it, create several 3.2 backends, and then XFER accounts away from the 2.4 backend to the new backends.
If I understand the issue correctly, the 2.4 frontend would currently not detect the 3.2 backends correctly, causing an index downgrade when XFERing from 2.4 to 3.2. Is that correct?

**Comment by elliefm** [comment]
That's exactly correct.  All the index data that wasn't available in like Cyrus 2.2 (index version 6) would be stripped out during the XFER, and your 3.2 deployment would end up missing a lot of information.  I'm not exactly sure what fields those are off the top of my head, but I suspect it would look a lot like data loss to users accessing the new backend.

It would also definitely lose GUIDs, which means you wouldn't be able to replicate from those new backends to elsewhere until you had reconstructed every single mailbox to recalculate them, which would be very expensive because it'd require calculating a sha1 sum of every single message on the entire store...

With the new patches, the XFER will recognise that 3.2 is newer than 2.4, and will send everything it's got.  You will still want to reconstruct -Vmax those mailboxes eventually, to get the most out of 3.2, but at least you won't be losing features/data in the process.

**Comment by hagedose** [comment]
Thanks for the clarification. If you decide against a new release for 2.4, should I try to apply all the commits in #3136?

**Comment by elliefm** [comment]
Good question.  Nah, that PR also has a master-only refactor in it, so none of it will patch cleanly onto the older releases.

But!  I've already done the work of backporting the important bits to the stable branches.  :)  For 2.4, they are (in order to apply them, i.e. the reverse of what git log shows): 

* b6a68e9003 imapd: fix bug in 2.3.x version detection
* 356e5bc6d9 imapd: recognise remote Cyrus versions 3.1, 3.2, 3.3
* 21db86c91c imapd: syslog when downgrading index to v6

But you could get away with just 356e5bc6d9 if you wanted to make the absolute minimum necessary change.

For anyone else reading along: functionally-identical commits to these three are also already on the 2.5, 3.0, and 3.2 branches, but have different commit id's because they have different parents.  So, if you need these fixes before a release is ready, check the cyrus-imapd-[version] branch for your install and grab the relevant commits.

**Comment by elliefm** [comment]
2.4.21, 2.5.16, 3.0.14 and 3.2.3 are all now available containing these fixes
